go/mysql/flavor.go

```diff
@@ -0,0 +1,258 @@
+	// returns NULL if GTIDs are not enabled.
+	waitUntilPositionCommand(ctx context.Context, pos Position) (string, error)
+
+	// makeBinlogEvent takes a raw packet from the MySQL binlog
+	// stream connection and returns a BinlogEvent through which
+	// the packet can be examined.
+	makeBinlogEvent(buf []byte) BinlogEvent
+
+	// enableBinlogPlaybackCommand and disableBinlogPlaybackCommand return an
+	// optional command to run to enable or disable binlog
+	// playback. This is used internally in Google, as the
+	// timestamp cannot be set by regular clients.
+	enableBinlogPlaybackCommand() string
+	disableBinlogPlaybackCommand() string
+}
+
+// mariaDBReplicationHackPrefix is the prefix of a version for MariaDB 10.0
+// versions, to work around replication bugs.
+const mariaDBReplicationHackPrefix = "5.5.5-"
+
+// mariaDBVersionString is present in
+const mariaDBVersionString = "MariaDB"
+
+// fillFlavor fills in c.Flavor based on c.ServerVersion.
+// This is the same logic as the ConnectorJ java client. We try to recognize
+// MariaDB as much as we can, but default to MySQL.
+//
+// MariaDB note: the server version returned here might look like:
+// 5.5.5-10.0.21-MariaDB-...
+// If that is the case, we are removing the 5.5.5- prefix.
+// Note on such servers, 'select version()' would return 10.0.21-MariaDB-...
+// as well (not matching what c.ServerVersion is, but matching after we remove
+// the prefix).
+func (c *Conn) fillFlavor() {
+	if strings.HasPrefix(c.ServerVersion, mariaDBReplicationHackPrefix) {
+		c.ServerVersion = c.ServerVersion[len(mariaDBReplicationHackPrefix):]
+		c.flavor = mariadbFlavor{}
+		return
+	}
+
+	if strings.Index(c.ServerVersion, mariaDBVersionString) != -1 {
+		c.flavor = mariadbFlavor{}
+		return
+	}
+
+	c.flavor = mysqlFlavor{}
+}
+
+//
+// The following methods are dependent on the flavor.
+// Only valid for client connections (will panic for server connections).
+//
+
+// IsMariaDB returns true iff the other side of the client connection
+// is identified as MariaDB. Most applications should not care, but
+// this is useful in tests.
+func (c *Conn) IsMariaDB() bool {
+	_, ok := c.flavor.(mariadbFlavor)
+	return ok
+}
+
+// MasterPosition returns the current master replication position.
+func (c *Conn) MasterPosition() (Position, error) {
+	gtidSet, err := c.flavor.masterGTIDSet(c)
+	if err != nil {
+		return Position{}, err
+	}
+	return Position{
+		GTIDSet: gtidSet,
+	}, nil
+}
+
+// SendBinlogDumpCommand sends the flavor-specific version of
+// the COM_BINLOG_DUMP command to start dumping raw binlog
+// events over a slave connection, starting at a given GTID.
+func (c *Conn) SendBinlogDumpCommand(slaveID uint32, startPos Position) error {
+	return c.flavor.sendBinlogDumpCommand(c, slaveID, startPos)
+}
+
+// ResetReplicationCommands returns the commands to completely reset
+// replication on the host.
+func (c *Conn) ResetReplicationCommands() []string {
```

Comment:
One worry about this approach is that the Conn API may get crowded due to these flavor-specific abstractions. But this is still better than what we had before.

In the long run, we can consider supporting these through vitess-specific SQL statements (like `RESET REPLICATION`) that get translated into flavor-specific commands under the covers.